tools/OutlineService/OutlineService/OutlineService.cs

```diff
@@ -904,48 +933,62 @@ private void NetworkAddressChanged(object sender, EventArgs evt)
                 eventLog.WriteEntry($"network changed but no gateway found: {e.Message}");
             }
 
-            // Only send on change, to prevent duplicate notifications (mostly
-            // harmless but can make debugging harder). Note how we send the
-            // RECONNECTED event before we know all the updates have succeeded:
-            // cheating, but it alerts the user sooner *and* if we were able to
-            // connect in the first place then it's extremely likely we'll be
-            // able to reconnect.
-            if (previousGatewayIp != gatewayIp || previousGatewayInterfaceIndex != gatewayInterfaceIndex)
+            if (previousGatewayIp == gatewayIp && previousGatewayInterfaceIndex == gatewayInterfaceIndex)
             {
-
-                SendConnectionStatusChange(gatewayIp == null ? ConnectionStatus.Reconnecting : ConnectionStatus.Connected);
+                // Only send on actual change, to prevent duplicate notifications (mostly
+                // harmless but can make debugging harder).
+                eventLog.WriteEntry($"network changed but gateway and interface stayed the same");
+                return; 
             }
-
-            if (gatewayIp == null)
+            else if (gatewayIp == null)
             {
+                SendConnectionStatusChange(ConnectionStatus.Reconnecting);

```

Comment:
I don't understand this block.  if we're reconnecting, why do we stop routing and remove the redirect?  Also below, it seems that the code doesn't match the comment -- are we adding or removing the loopback redirect here?

Reply:
`StopRoutingIpv4` redirects traffic to the loopback interface.
`RemoveIpv4Redirect` stops capturing IPv4 traffic by deleting the ("IPv4 redirect") route to the TAP device .

Reply:
Maybe we can rename these functions since "redirect" is now overloaded -- some sort of names referenceing the TAP device and the loopback interface?  Maybe `Add/RemoveIpv4LoopbackRedirect` and `Add/RemoveIpv4TapRoute`?

Reply:
Good suggestion. I renamed the TAP route functions in 2293b80 and re-signed the binary.